enclave-runner/src/usercalls/mod.rs

```diff
@@ -82,143 +107,353 @@ impl<W: Write> Write for WriteOnly<W> {
-    fn read(&mut self, buf: &mut [u8]) -> IoResult<usize> {
-        self.0.lock().read(buf)
+impl<T> AsyncWrite for WriteOnly<T>
+where
+    T: std::io::Write,
+{
+    fn shutdown(&mut self) -> Poll<(), std::io::Error> {
+        Ok(().into())
     }
 }
 
-impl<'a, T: SharedStream<'a>> Write for &'a Shared<T> {
-    fn write(&mut self, buf: &[u8]) -> IoResult<usize> {
-        self.0.lock().write(buf)
+struct Stdin;
+
+impl Read for Stdin {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        const BUF_SIZE: usize = 8192;
+
+        trait AsIoResult<T> {
+            fn as_io_result(self) -> io::Result<T>;
+        }
+
+        impl<T> AsIoResult<T> for Async<T> {
+            fn as_io_result(self) -> io::Result<T> {
+                match self {
+                    Async::Ready(v) => Ok(v),
+                    Async::NotReady => Err(io::ErrorKind::WouldBlock.into()),
+                }
+            }
+        }
+
+        struct AsyncStdin {
+            rx: async_mpsc::Receiver<VecDeque<u8>>,
+            buf: VecDeque<u8>,
+        }
+
+        lazy_static::lazy_static! {
+            static ref STDIN: Lock<AsyncStdin> = {
+                let (mut tx, rx) = async_mpsc::channel(8);
+                thread::spawn(move || {
+                    let mut buf = [0u8; BUF_SIZE];
+                    while let Ok(len) = io::stdin().read(&mut buf) {
+                        if len == 0 {
+                            continue
+                        }
+
+                        if tx.try_send(buf[..len].to_vec().into()).is_err() {
+                            return
+                        };
+                    }
+                });
+                Lock::new(AsyncStdin { rx, buf: VecDeque::new() })
+            };
+        }
+
+        let mut stdin = STDIN.clone().poll_lock().as_io_result()?;
+        if stdin.buf.is_empty() {
+            let pipeerr = || -> io::Error { io::ErrorKind::BrokenPipe.into() };
+            stdin.buf = stdin
+                .rx
+                .poll()
+                .map_err(|_| pipeerr())?
+                .as_io_result()?
+                .ok_or_else(pipeerr)?;
+        }
+        let inbuf = match stdin.buf.as_slices() {
+            (&[], inbuf) => inbuf,
+            (inbuf, _) => inbuf,
+        };
+        let len = cmp::min(buf.len(), inbuf.len());
+        buf[..len].copy_from_slice(&inbuf[..len]);
+        stdin.buf.drain(..len);
+        Ok(len)
     }
+}
 
-    fn flush(&mut self) -> IoResult<()> {
-        self.0.lock().flush()
+pub trait AsyncStream: AsyncRead + AsyncWrite + 'static + Send + Sync {
+    fn poll_read_alloc(&mut self) -> tokio::prelude::Poll<Vec<u8>, std::io::Error>
+    {
+        let mut v: Vec<u8> = vec![0; 8192];
+        self.poll_read(v.as_mut_slice()).map(move |size| {
+            match size {
+                Async::NotReady => Async::NotReady,
+                Async::Ready(size) => {
+                    v.resize(size, 0);
```

Comment:
What is the reason to use truncate over resize? I understand that truncate will not increase the size of the vector
Why do we want to limit the size to `8192`?

Reply:
Because `poll_read()` is only given a mutable slice (not the vec) which means it is limited to read up to 8K. Using truncate is better because we don't want to allocate unnecessarily in case the function returns  a number > 8K